src/rust/engine/options/src/parse.rs

```diff
@@ -55,23 +55,52 @@ peg::parser! {
             = "\"" s:double_quoted_character()* "\"" { s.into_iter().collect() }
 
         rule double_quoted_character() -> char
-            = quoted_character("\"")
-            / escaped_character()
+            = escaped_character() / quoted_character("\"")
 
         rule single_quoted_string() -> String
             = "'" s:single_quoted_character()* "'" { s.into_iter().collect() }
 
         rule single_quoted_character() -> char
-            = quoted_character("'")
-            / escaped_character()
+            = escaped_character() / quoted_character("'")
 
         // NB: ##method(X) is an undocumented peg feature expression that calls input.method(pos, X)
         // (see https://github.com/kevinmehall/rust-peg/issues/283).
         rule quoted_character(quote_char: &'static str) -> char
-            = !(##parse_string_literal(quote_char) / "\\") c:$([_]) { c.chars().next().unwrap() }
+            = !(##parse_string_literal(quote_char)) c:$([_]) { c.chars().next().unwrap() }
+
+        // Python string literal escape sequences.
+        // See https://docs.python.org/3/reference/lexical_analysis.html#escape-sequences.
+        // Note that only backslash, single-quote, linefeed, carriage return and horizontal tab
+        // are also Rust character escape sequences.
+
+        rule escaped_ba() -> char = "\\\\" { '\\' }
+        rule escaped_sq() -> char = "\\'" { '\'' }
+        rule escaped_dq() -> char = "\\\"" { '"' }
+        rule escaped_be() -> char = "\\a" { '\x07' }
+        rule escaped_bs() -> char = "\\b" { '\x08' }
+        rule escaped_ff() -> char = "\\f" { '\x0c' }
+        rule escaped_lf() -> char = "\\n" { '\n' }
+        rule escaped_cr() -> char = "\\r" { '\r' }
+        rule escaped_ht() -> char = "\\t" { '\t' }
+        rule escaped_vt() -> char = "\\v" { '\x0b' }
+
+        // Python octal escapes take 1-3 digits.
+        rule escaped_octal() -> char = "\\" s:$(['0'..='7']*<1,3>) {
+            char::from_u32(u32::from_str_radix(s, 8).unwrap()).unwrap()
+        }
+
+        // Python hex escapes take exactly 2 digits. Note that we mirror the Python behavior of
+        // always consuming the next two characters and failing if they aren't valid hex digits.
+        rule escaped_hex() -> char = "\\x" s:$([_] [_]) {
+            char::from_u32(u32::from_str_radix(s, 16).unwrap_or_else(
+                |_| panic!("expected two hex digits, found {}", s))).unwrap()
```

Comment:
Oh, sorry, I missed this the first time around... panicking for invalid user input is generally not the Rust convention (vs. returning `Err` and handling that upstream). Is that how we have to report errors here?

Reply:
AFAICT, yes.  There is special "conditional action" syntax (`{?` instead of `{`) in the peg parser macros to support a rule returning a `Result`. But returning an `Err` in that case just causes the rule not to match, rather then error out. And then the `non_escaped_character()` rule will match. So `\xZZ` will produce a literal `\xZZ` rather than an error. That's different than the current behavior, but maybe that's OK? Of course in that case the user won't see any error message anyway...

Reply:
I'll have a think on it, maybe there's a way around it. But probably only by changing the return type of the rule to String (so it can return `\xZZ`), and that might have annoying ripple effects

Reply:
OK, figured something out by explicitly disallowing `\x` to match `non_escaped_character()`.  PTAL!

Note: we are still using `unwrap()` (which is a panic) in various places. But in those other cases the panic can't be triggered by user input, only by programmer error.